# Metallicity Switchboard

In [3]:
import astropy.units as u
import astropy.constants as ac
from astropy.cosmology import FlatLambdaCDM
import numpy as np
import pandas as pd
import os
from os.path import expanduser
from astropy.table import Table
from glob import glob

In [4]:
# gala is a data table of a Final_galinfo.csv file which 
# is subset of data from the galaxyinfo.xlsx data from CGM^2

galaxy = pd.read_csv('Final_galinfo.csv')

galaxy.index += 2   # This is so the subset of data matches the exact 
                  # index value of the galaxyinfo.xlsx data

galaxy.to_numpy()   # This turns the DataFrame into a NumPy array.

galaxy

specfile                       id      OBJECT  \
2    J022601.51+001532.5.fits  J022601.51+001532.5._03  J0226+0015   
3    J022602.41+001536.6.fits  J022602.41+001536.6._01  J0226+0015   
4    J022606.98+001451.4.fits  J022606.98+001451.4._02  J0226+0015   
5    J022607.08+001612.7.fits  J022607.08+001612.7._01  J0226+0015   
6    J022608.02+001427.2.fits  J022608.02+001427.2._03  J0226+0015   
..                        ...                      ...         ...   
779  J234501.13-010020.1.fits  J234501.13-010020.1._04  J2345-0059   
780  J234502.05-005738.7.fits  J234502.05-005738.7._04  J2345-0059   
781  J234503.08-005707.5.fits  J234503.08-005707.5._04  J2345-0059   
782  J234508.21-005959.2.fits  J234508.21-005959.2._04  J2345-0059   
783  J234508.25-005921.1.fits  J234508.25-005921.1._04  J2345-0059   

               cgm_name      gal_ra  gal_dec         z      OII_flux  \
2    J0226+0015_271_194   36.506302  0.25904  0.352715 -9.900000e+01   
3    J0226+0015_272_181   36.510059  0.26017  0.297832 -9.900000e+01   
4    J0226+0015_251_119   36.529079  0.24762  0.384662  4.200000e-17   
5    J0226+0015_291_119   36.529518  0.27020  0.299556 -9.900000e+01   
6    J0226+0015_237_115   36.533401  0.24088  0.506584  4.100000e-17   
..                  ...         ...      ...       ...           ...   
779   J2345-0059_167_45  356.254730 -1.00558  0.454484  9.050000e-17   
780   J2345-0059_12_120  356.258514 -0.96074  0.311012  6.920000e-17   
781   J2345-0059_15_154  356.262817 -0.95208  0.428484  1.050000e-16   
782  J2345-0059_101_119  356.284180 -0.99979  0.663357  3.600000e-17   
783   J2345-0059_83_118  356.284363 -0.98920  0.305693 -9.900000e+01   

     OII_flux_sig  OII_LQ  ...      NII_flux  NII_flux_sig  NII_LQ  \
2   -9.900000e+01   -99.0  ...  1.120000e-17  2.570000e-18     1.0   
3   -9.900000e+01   -99.0  ... -9.900000e+01 -9.900000e+01   -99.0   
4    5.280000e-18     1.0  ...  5.910000e-17  9.820000e-19     1.0   
5   -9.900000e+01   -99.0  ...  3.970000e-18  7.050000e-19     1.0   
6    3.290000e-18     1.0  ... -9.900000e+01 -9.900000e+01   -99.0   
..            ...     ...  ...           ...           ...     ...   
779  3.730000e-18     1.0  ... -9.900000e+01 -9.900000e+01     0.0   
780  8.390000e-18     1.0  ... -9.900000e+01 -9.900000e+01     0.0   
781  5.920000e-18     1.0  ... -9.900000e+01 -9.900000e+01     0.0   
782  4.320000e-18     1.0  ... -9.900000e+01 -9.900000e+01     0.0   
783 -9.900000e+01     0.0  ... -9.900000e+01 -9.900000e+01     0.0   

        SIIa_flux  SIIa_flux_sig  SIIa_LQ     SIIb_flux  SIIb_flux_sig  \
2    1.290000e-17   7.990000e-19      1.0  1.760000e-18   1.140000e-18   
3   -9.900000e+01  -9.900000e+01      2.0 -9.900000e+01  -9.900000e+01   
4   -9.900000e+01  -9.900000e+01    -99.0 -9.900000e+01  -9.900000e+01   
5    6.710000e-18   7.030000e-19      1.0  4.400000e-18   7.560000e-19   
6   -9.900000e+01  -9.900000e+01    -99.0 -9.900000e+01  -9.900000e+01   
..            ...            ...      ...           ...            ...   
779 -9.900000e+01  -9.900000e+01      0.0 -9.900000e+01  -9.900000e+01   
780 -9.900000e+01  -9.900000e+01      0.0 -9.900000e+01  -9.900000e+01   
781 -9.900000e+01  -9.900000e+01      0.0 -9.900000e+01  -9.900000e+01   
782 -9.900000e+01  -9.900000e+01      0.0 -9.900000e+01  -9.900000e+01   
783 -9.900000e+01  -9.900000e+01      0.0 -9.900000e+01  -9.900000e+01   

     SIIb_LQ  Comment  
2        1.0      NaN  
3        2.0      NaN  
4      -99.0      NaN  
5        1.0      NaN  
6      -99.0      NaN  
..       ...      ...  
779      0.0      NaN  
780      0.0      NaN  
781      0.0      NaN  
782      0.0      NaN  
783      0.0      NaN  

[782 rows x 29 columns]

### This notebook is mainly a framework. It would be ideal if your individual functions returned a NumPy array that is the same length as 'galaxy' (n=782), indexed correctly, values that can be calculated are dtype=float, values that cannot be calculated are (np.nan). The bold print cells below represent functions that need to be written.


## Calculate 'O3HB'

-Add column 'O3HB' to 'galaxy'

## Calculate 'N2' 

-Add column 'N2' to 'galaxy'

## Calculate 'O3N2'

-Add 'O3N2' column to 'galaxy'

-Check 'O3N2' values

  IF O3N2 < 1.9

## Calculate 'Z_O3N2-'

ELSEIF O3N2 > 1.9

## Calculate 'Z_O3N2+'

-If 'O3N2' or 'O3HB' could not be calculated (NaN) and we only have N2 index then the equation for calculating 'Z_O3N2+' (when O3N2 > 1.9) can be used for calculating metallicity.

-If 'O3N2' or 'N2' could not be calculated (NaN) and we only have O3HB values:

*If there is no OII flux, then metallicity cannot be calculated, we're done.

*If there is OII flux then we need to use the R23 method

## Calculate R23

-From the conversation in the the last meeting (and please correct me if I'm wrong) I was under the impression that we are only going to use the R23 'Lower Branch' calculations only.

-Create a 'Flag_Z' column

-Save to .csv